### PR TITLE
HDDS-6735. Provide configuration for enabling OMLockMetrics

### PR DESCRIPTION
**Pull request by tanvipenumudy**
## What changes were proposed in this pull request?

To introduce a configuration: `ozone.om.lock.metrics.collection.enable` (default value = `false`) to disable `OMLockMetrics` related code collection by default.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6735

## How was this patch tested?

Existing unit and integration test-cases.

### CODE REVIEW

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneManagerLock.java [hunk 1]
```diff
@@ -202,6 +196,26 @@ private boolean lock(Resource resource, String resourceName,
     }
   }
 
+  private void updateLockMetrics(Resource resource, String resourceName,
```
Comment (rakeshadr) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneManagerLock.java:199]: Please rename the method as its doing actual locking as well as updating metrics.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneManagerLock.java [hunk 2]
```diff
@@ -202,6 +196,26 @@ private boolean lock(Resource resource, String resourceName,
...
+                         Consumer<String> lockFn, String lockType) {
+    long startWaitingTimeNanos = Time.monotonicNowNanos();
+    lockFn.accept(resourceName);
+
+    /**
+     *  read/write lock hold count helps in metrics updation only once in case
+     *  of reentrant locks.
+     */
+    if (lockType.equals(READ_LOCK) &&
```
Comment (rakeshadr) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneManagerLock.java:208]: Can we make this more readable, something like to use a switch case:

```
if (omLockMetricsCollectionEnabled) {
  switch (lockType) {
  case READ_LOCK:
    updateReadLockMetrics(resource, resourceName, startWaitingTimeNanos);
    break;
  case WRITE_LOCK:
    updateWriteLockMetrics(resource, resourceName, startWaitingTimeNanos);
    break;
  default:
    break;
  }
} else {
  lockFn.accept(resourceName);
}
```

File: hadoop-hdds/common/src/main/resources/ozone-default.xml [hunk 3]
```diff
@@ -2866,6 +2866,15 @@
     </description>
   </property>
 
+  <property>
+    <name>ozone.om.lock.metrics.collection.enable</name>
+    <tag>OZONE, OM</tag>
+    <value>false</value>
+    <description>Defaults to false. If true, the OMLockMetrics collection is
+      enabled. If false, it is disabled.
```
Comment (kerneltime) [inline hadoop-hdds/common/src/main/resources/ozone-default.xml:2874]:
```suggestion
    <description>If set to true enables OMLockMetrics collection. Defaults to false/disabled.
```

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneManagerLock.java [hunk 4]
```diff
@@ -176,23 +182,11 @@ private boolean lock(Resource resource, String resourceName,
...
-        updateReadLockMetrics(resource, startWaitingTimeNanos);
-      }
-      if (lockType.equals(WRITE_LOCK) &&
-          (manager.getWriteHoldCount(resourceName) == 1) &&
-          manager.isWriteLockedByCurrentThread(resourceName)) {
-        updateWriteLockMetrics(resource, startWaitingTimeNanos);
+      if (!omLockMetricsCollectionEnabled) {
+        lockFn.accept(resourceName);
+      } else {
+        updateLockMetrics(resource, resourceName, lockFn, lockType);
```
Comment (kerneltime) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneManagerLock.java:188]: This code needs jumping around to figure out what is the exact set of actions that will take place.
I would recommend, adding a wrapper that actually implements the check for if lock metrics should be collected. The code here should just be lockFn.accept -> measureAndAccept(lockFn, resource, resourceName, lockType). In the the wrapper check if the flag is set.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneManagerLock.java (hunk 2 above)
Comment (tanvipenumudy) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneManagerLock.java]: Thank you for the suggestion, I have made the required changes to incorporate the switch case.

File: hadoop-hdds/common/src/main/resources/ozone-default.xml (hunk 3 above)
Comment (tanvipenumudy) [reply hadoop-hdds/common/src/main/resources/ozone-default.xml]: Thank you for the review, I have made the changes!

**Pull request closed without merging**